Public/Src/Engine/Scheduler/Artifacts/FileContentManager.cs

```diff
@@ -1579,9 +1579,12 @@ private void PopulateArtifactsToMaterialize(PipArtifactsState state, bool? allow
 
                     if (sealDirectoryKind == SealDirectoryKind.Opaque)
                     {
-                        // Dynamic directories must be deleted before materializing files
-                        // We don't want this to happen for shared dynamic ones
-                        AddDirectoryDeletion(state, artifact.DirectoryArtifact);
+                        if (Configuration.Sandbox.UnsafeSandboxConfiguration.PreserveOutputs != PreserveOutputsMode.Enabled || !PipArtifacts.IsPreservedOutputByPip(state.PipInfo.UnderlyingPip, directory.Path, Context.PathTable))

```

Comment:
nit: rhs of `||` on new line for readability..

Reply:
i cannot update Mike's branch. As the rest of the changes looks good, I will merge it.